### DISCUSSION TITLE
No calibration / boundary limit  needed?

### DISCUSSION THREAD

**Discussion by holger8080** [doudar/SmartSpin2k#159] [Q&A] [answered]
Thanks for the project, looks really cool. I'm about to order parts to build one. 

Maybe my question is trivial, but I'm wondering: why do I not see any kind of calibration anywhere?
Can't it happen that the motor tries to go beyond the resistance knob limit? That means that typically:
- The motor or the PCB would break in one form or another (overheating etc)
- The gears or adapter would break

So e.g. when there's a simulated downhill, the stepper motor tries to go below the mechanical resistance limit of the bike and something bad happens?
Don't I need to specify upper and lower limits? Or have a mechanical switch installed, like 3d-printers have for the axes?

I haven't seen any calibration in the wiki nor at a quick glance of the code. Did I miss something?

**Comment by fireyeti** [comment]
@holger8080

This is a good issue to add to the wiki. In my experience the range of the resistance knob of my bike and many I have used is much greater than what the SS2K will request even with the incline multiplier on maximum.

The stepper motor with the power all the way at 2000 will stop turning if it encounters too much resistance. I can get to around 800W on my resistance type spin bike (leather pad rubbing on metal flywheel) before the stepper no longer has enough power to rotate the resistance knob. This is by design to prevent the plastic gears from stripping or any internal parts breaking.

**Reply by holger8080** [reply]
Cool thanks. And I'm not so much worried about hitting the 1000W+ limit than hitting the lower limit. ;-)

**Comment by doudar** [comment, marked as answer]
> Thanks for the project, looks really cool. I'm about to order parts to build one.
> 
> Maybe my question is trivial, but I'm wondering: why do I not see any kind of calibration anywhere?
> Can't it happen that the motor tries to go beyond the resistance knob limit? That means that typically:
> 
> * The motor or the PCB would break in one form or another (overheating etc)
> * The gears or adapter would break
> 
> So e.g. when there's a simulated downhill, the stepper motor tries to go below the mechanical resistance limit of the bike and something bad happens?
> Don't I need to specify upper and lower limits? Or have a mechanical switch installed, like 3d-printers have for the axes?
> 
> I haven't seen any calibration in the wiki nor at a quick glance of the code. Did I miss something?

As @fireyeti mentioned, we're currently relying on the unit to be "self adapting" to it's installation without requiring homing or anything else. It does this by simply being torque limited. As there's no issue with stepper motors slipping it works pretty good.

That said, eventually we desire some type of homing routine. The current stepper driver supports this without mechanical end stops as it allows us to read actual current and detect missed steps. Homing is on the road map but it's a ways out. If anyone would like to take a whack at it feel free! Marlin probably has some code in it that we could base off of as they allow sensor less homing. The benefit we have is that we don't currently have to worry about breaking anything if hit the mechanical limits whereas Marlin does.      

FMI:
https://github.com/MarlinFirmware/Marlin
https://github.com/MarlinFirmware/Marlin/blob/2.0.x/Marlin/src/module/endstops.cpp (search for "tmc")
https://github.com/MarlinFirmware/Marlin/blob/e3a12c3c28c2b6c4882eaa6549c6053b4d38953e/Marlin/src/module/motion.cpp#L1144

**Reply by holger8080** [reply]
Perfect, thanks. If it has a torque limit, that will do the job for me.

**Comment by drz01** [comment]
Homing sounds awesome! Currently loud noises from the stepper when it hits the resistance limits, since I had to up the shift step. This is on lowish torque too, with a 19V power supply and the original gear set.
